[DOC] Update metrics generator doc and add best practices (#2563)
grafana/tempo: docs/sources/tempo/metrics-generator/active-series.md

---
aliases:
- /docs/tempo/latest/metrics-generator/active-series
title: Active series
menuTitle: Active series
description: Learn about active series and how they are calculated.
weight: 100
---

# Active series

Metrics generated by Tempo's metrics generator provide an easy route to obtaining both RED (Rate/Error/Duration) metrics as well as interdependency graphs between services in a trace (the Service Graph functionality in Grafana).
These capabilities rely on a set of generated span metrics and service metrics.

> Any spans that might be ingested by Tempo could have 16 metrics. However, this doesn't mean that every time a span is ingested that a new active series count is created.

Author: I like your wording better than what we had. I wrote 16, but I probably counted wrong. :) Jennie P added two new metrics, job and instance, recently, to the span metrics (see doc below). I'm not sure when these metrics will be released (i.e., part of 2.2) so best to probably not include them for now.

This means that when the span completes `status_code` is now either `STATUS_CODE_OK` or `STATUS_CODE_ERROR`.
Because of that, the label values can be one of two values on a metric, and we now have two active series being generated based on the `status_code`, one for the `OK` status and one for the error.

Again, we could loop once, 10 times, 100 times... but there will only ever be two active series.

If we now looked at Prometheus instant values for `traces_spanmetrics_call_total` we'd now see the following table:

| service   | span_name | span_kind          | status_code       | Metric value |
| --------- | --------- | ------------------ | ----------------- | ------------ |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_OK    | 96           |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_ERROR | 24           |

What happens if we call out to another service though? Let's add an option where, based on some arbitrary data, we sometimes call out to another service, but otherwise just stay locked within our own service:

![Multiple spans with loops and errors](/static/img/docs/tempo/SingleSpanLoopErrorAnotherService.jpeg)

> Here, `span1`'s `span_kind` label would now be one of either `SPAN_KIND_INTERNAL` or `SPAN_KIND_CLIENT` (as it has acted as a client calling another service).

Reviewer: ```suggestion
In this scenario, `span1`'s `span_kind` label would now be one of either `SPAN_KIND_INTERNAL` or `SPAN_KIND_CLIENT` (as it has acted as a client calling a downstream server).
```

If a call to the downstream service could also potentially fail, then for `SPAN_KIND_CLIENT`, the `status_code` could be either `STATUS_CODE_ERROR` or `STATUS_CODE_OK`.

At this point, `traces_spanmetrics_call_total` would have four different variations in labels:

| service   | span_name | span_kind          | status_code       | Metric value |
| --------- | --------- | ------------------ | ----------------- | ------------ |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_OK    | 34           |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_ERROR | 6            |
| Service 1 | span1     | SPAN_KIND_CLIENT   | STATUS_CODE_OK    | 23           |
| Service 1 | span1     | SPAN_KIND_CLIENT   | STATUS_CODE_ERROR | 3            |

Because of the variation in values, we now have four active series for our metric instead of one. But, as far as Service 1 is concerned, there's still only four active series, because there isn't any other variation of the values for labels. You can run 1 trace, 10 traces, 100 traces (each with however many loops of spans there are) and only four active series will ever be produced.

> In reality, we've only told half the story in our last diagram. We called a second service, _Service 2_, which continues the trace by adding a new span, `span2`.

Reviewer: ```suggestion
We've actually only told half the story in our last diagram. _Service 1_ called a second service, _Service 2_, which continues the trace by adding a new span, `span2`.
```

| --------- | --------- | ------------------ | ----------------- | ------------ |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_OK    | 89           |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_ERROR | 13           |
| Service 1 | span1     | SPAN_KIND_CLIENT   | STATUS_CODE_OK    | 44           |
| Service 1 | span1     | SPAN_KIND_CLIENT   | STATUS_CODE_ERROR | 9            |
| Service 2 | span2     | SPAN_KIND_SERVER   | STATUS_CODE_OK    | 30           |
| Service 2 | span2     | SPAN_KIND_SERVER   | STATUS_CODE_ERROR | 14           |
| Service 2 | span2     | SPAN_KIND_INTERNAL | STATUS_CODE_OK    | 99           |
| Service 2 | span2     | SPAN_KIND_INTERNAL | STATUS_CODE_ERROR | 23           |

At this point, all our traces will be composed of two potential span names, each of which produce two separate types of `span_kind` and two separate types of `status_code`. So we have eight active series for a metric.

The variability of values for each potential span condition determines the number of active series being produced by Tempo when ingesting spans for a trace, and not the number of traces of spans that are seen.

## Custom span attributes

> Finally, there's one other consideration for active series: extra label key/value pairs that can be added onto metrics.

Reviewer: ```suggestion
There's another consideration for active series: extra label key/value pairs that can be added onto metrics from a span's attributes.
```

The Tempo metrics generator allows the user to use arbitrary span attributes to be created as label pairs for metrics.
When considering the number of active series generated, you also need to determine how many possible values there are for the span attribute being turned into a label.

> For example, if someone added an `http.method` span attribute into a metric label pair, there are five possible values (because there are five REST methods):

Reviewer: ```suggestion
For example, if you added an `http.method` span attribute into a metric label pair, there are five possible values (because there are five possible REST methods):
```

- `HEAD`
- `GET`
- `POST`
- `PUT`
- `DELETE`

> If this label pair is added to every span metric, that's another possible 5 active series generated for each metric (in all likelihood this is a very worst case scenario, very few spans will call all five REST methods), so instead of 8 active series in the last table above, we'd have 40 (8 * 5).

Reviewer: ```suggestion
If this label pair is added to every span metric, that's another 5 *potential*  active series generated for each metric (in all likelihood this is a very worst case scenario, very few spans will call all five REST methods), so instead of 8 active series in the last table above, we'd have 40 (8 * 5).
```

---
aliases:
- /docs/tempo/latest/metrics-generator/active-series
title: Active series
menuTitle: Active series
description: Learn about active series and how they are calculated.
weight: 100
---

> # Active series

Author: Thank you! That definition is great, plus we can link to that doc in case people want additional information.

> Metrics generated by Tempo's metrics generator provide an easy route to obtaining both RED (Rate/Error/Duration) metrics as well as interdependency graphs between services in a trace (the Service Graph functionality in Grafana).
> These capabilities rely on a set of generated span metrics and service metrics.
> 
> Any spans that are ingested by Tempo could potentially create *up* to 13 metrics. However, this doesn't mean that every time a span is ingested that a new active series is created.
> 
> The number of active series generated depends on the label pairs generated from span data that are associated with the metrics (just as any other Prometheus format data).

Author: ```suggestion

An active series is a time series that receives new data points or samples. When you stop writing new datapoints to a time series, shortly afterwards it is no longer considered active. 

Metrics generated by Tempo's metrics generator can provide both RED (Rate/Error/Duration) metrics and interdependency graphs between services in a trace (the Service Graph functionality in Grafana).
These capabilities rely on a set of generated span metrics and service metrics.

Any spans that are ingested by Tempo could potentially create up to 13 metrics. However, this doesn't mean that every time a span is ingested that a new active series is created.

The number of active series generated depends on the label pairs generated from span data that are associated with the metrics (just as any other Prometheus format data).

For additional information, refer to the [Active series](/docs/grafana-cloud/billing-and-usage/active-series-and-dpm/#active-series) documentation. 
```

![Single span with loop](/static/img/docs/tempo/SingleSpanLoop.jpeg)

If we looked at the Prometheus data, we'd see an instant value for `traces_spanmetrics_call_total` similar to the table. Again, one active series for the metric:

| service   | span_name | span_kind          | status_code    | Metric value |
| --------- | --------- | ------------------ | -------------- | ------------ |
| Service 1 | span1     | SPAN_KIND_INTERNAL | STATUS_CODE_OK | 120          |

However, let's now assume that it does loop and there are occasionally errors:

![Single span with loop and errors](/static/img/docs/tempo/SinglespanLoopError.jpeg)

There are now two potential outcomes for a span when the code loops: one where everything successfully completes and one where there is an error.
This means that when the span completes `status_code` is now either `STATUS_CODE_OK` or `STATUS_CODE_ERROR`.
Because of that, the label values can be one of two values on a metric, and we now have two active series being generated based on the `status_code`, one for the `OK` status and one for the error.

> Again, we could loop once, 10 times, 100 times... but there will only ever be two active series.

Author: ```suggestion
Again, we could loop once, 10 times, 100, or more times, but there will only ever be two active series.
```
